### PR TITLE
"Config" attribute on command-line properties to allow default values to be set in app configuration file.

### PR DESCRIPTION
**Pull request by jazmatician**
This is some functionality I added to BizArk to allow you to mark command-line attributes as manageable via the system's configuration functionality. I use ConfigurationManager, but there's an interface, IConfigurationProvider (different from Microsoft.Extensions.Configuration.IConfigurationProvider) that allows unit testing and other configuration systems (like the .Net Core approach). Started to implement the MS interface, but the TryGet/out paradigm is a pain to mock.
Sorry for all the extra edit lines in the ConsoleApp class, I didn't notice it reformatting the lines until I was making this pull request.
Also, this makes the application class itself a string-indexed object (i.e `this[string key]`), so you can reference the parameters by key. I needed this originally when I wanted to pass a configuration object to the logic class, but it just made doing the configuration so easy.

### CODE REVIEW

**Comment by bbrewder** [comment]
Sorry it took so long to take a look at this (busy with work and personal projects). I think I understand what you are doing and it seems reasonable. 

I'll likely accept a this pull request if you can do the following things:

1. Fix the formatting issues. It's difficult to see what's changed when everything appears changed,
2. Change the p variable name to "prop" and a variable name to "att". These are the names I use in other places in BizArk and I would like to be consistent. This isn't a deal breaker, just my preference - if you don't, I probably will change it.
3. Move the constructors to the top of the class, followed by fields. I typically order my classes with constructors, destructors, Dispose, constants, fields, properties, methods, event handlers, events. Again, not required, just preferred.
4. Write some basic documentation on how to use this feature. No need to be fancy, I can re-write it for the BizArk docs, I just want to make sure I understand it well enough to be able to write those docs (I don't want to guess, and if I have to figure it out on my own, this will take a lot longer to approve :) ). The docs should include anything the app developer should do to make use of it as well as anything the end user of the app can do (I think this is more of an end-user feature, but not sure).

File: BizArk.ConsoleApp/ConsoleApp.cs [hunk 1]
```diff
@@ -1,74 +1,141 @@
...
+
+        /// <summary>
+        /// If an alternate provider is required (Net Core's ConfigurationBuilder, for example, or for unit testing) 
+        /// this constructor can be used. 
+        /// </summary>
+        /// <param name="config">Configuration Provider to be used (</param>
+        public BaseConsoleApp(IConfigurationProvider config)
+        {
+            LogProps = TypeDescriptor.GetProperties(this.GetType());
+            foreach (var p in this.GetType().GetProperties())
```
Comment (bbrewder) [inline BizArk.ConsoleApp/ConsoleApp.cs:87]: Change p to prop.

File: BizArk.ConsoleApp/ConsoleApp.cs [hunk 2]
```diff
@@ -1,74 +1,141 @@
...
+            {
+                foreach (var a in p.GetCustomAttributes(true))
```
Comment (bbrewder) [inline BizArk.ConsoleApp/ConsoleApp.cs:89]: Change a to att.

File: BizArk.ConsoleApp/ConsoleApp.cs [hunk 3]
```diff
@@ -1,74 +1,141 @@
...
+        /// <returns></returns>
+        bool Error(Exception ex);
+    }
+
+    /// <summary>
+    /// A base class that can be used so developers don't have to implement all the methods.
+    /// </summary>
+    public abstract class BaseConsoleApp : IConsoleApp, IStringIndexedObject
+    {
+        private PropertyDescriptorCollection LogProps;
```
Comment (bbrewder) [inline BizArk.ConsoleApp/ConsoleApp.cs:42]: Change LogProps to mLogProps. Move to just under constructors.

File: BizArk.ConsoleApp/ConsoleApp.cs [hunk 4]
```diff
@@ -1,74 +1,141 @@
...
+        /// <returns></returns>
+        public virtual bool Error(Exception ex)
+        {
+            return false;
+        }
+
+        /// <summary>
+        /// Default constructor will pull configurations from the traditional .NET Framework configuration provider.
+        /// </summary>
+        public BaseConsoleApp() : this(new NetFrameworkConfigProvider()) { }
```
Comment (bbrewder) [inline BizArk.ConsoleApp/ConsoleApp.cs:77]: Move constructors to top of class.

**Comment by jazmatician** [comment]
Thanks so much for the detailed comments, I do a lot of code reviews, so I
know this took some time on your part to give it to me in an organized
punchlist. I'll work on implementing all of these suggestions, but the rest
of this month will be a little stacked up for me, so look for it in August
at the earliest. I think I'll try to get the docs done first, since that's
the part that would make me strictly redundant if that comes to it.
For now, reject the pull request and I'll work on a nice clean patch on a
new branch.

Thanks again,
Michael

**Comment by bbrewder** [comment]
Sounds good. I've got a bunch of projects keeping me busy this month (and probably next) anyway.

**Pull request closed without merging**